### DISCUSSION TITLE
How do I find all available versions for a package

### DISCUSSION THREAD

**Discussion by henrysachs** [wolfi-dev/os#3070] [Q&A] [answered]
I want to install the prometheus package but I only get the "latest" version. Is there a way to get all available versions for this specific package?

**Comment by CloudMass** [comment]
@henrysachs You can see the tags in the [Chainguard image catalog](https://console.enforce.dev/images/catalog?filter=all).
But from what I understand you will not be able to pull them unless you pay for "Standard Catalog" or "Custom Catalog" https://www.chainguard.dev/unchained/scaling-chainguard-images-with-a-growing-catalog-and-proactive-security-updates

![image](https://github.com/wolfi-dev/os/assets/6693421/ba439b30-84e8-4ab7-8e10-a9fb33f06846)

**Reply by henrysachs** [reply]
@CloudMass  I thought the apk packages would have different versions than the docker images and that I would be able to pull them without a subscription?

**Reply by CloudMass** [reply]
Oh, I misunderstood. I am not sure how to find the apk package version but my bet is that it's the same as the image tags.

**Reply by henrysachs** [reply]
Yeah I thought so too but unfortunately when using it in an apko file I get an error :(

**Reply by henrysachs** [reply]
@luhring @dlorenc Maybe you 2 could help me with this question? I see you guys all over the chainguard and wolfi place. 😅

**Reply by dlorenc** [reply, marked as answer]
You can try this:

```Dockerfile
$ apk update
$ apk search -a prometheus
newrelic-prometheus-configurator-1.4.0-r0
newrelic-prometheus-configurator-1.4.0-r1
newrelic-prometheus-configurator-1.4.1-r0
newrelic-prometheus-configurator-1.4.1-r1
newrelic-prometheus-configurator-1.4.2-r0
nri-prometheus-2.18.0-r0
nri-prometheus-2.18.1-r0
nri-prometheus-2.18.1-r1
nri-prometheus-2.18.1-r2
nri-prometheus-2.18.4-r0
prometheus-2.41.0-r0
prometheus-2.41.0-r1
prometheus-2.41.0-r2
prometheus-2.42.0-r0
prometheus-2.42.0-r1
prometheus-2.42.0-r2
prometheus-2.42.0-r3
prometheus-2.43.0-r0
prometheus-2.43.0-r1
prometheus-2.43.0-r2
prometheus-2.43.1-r0
prometheus-2.44.0-r0
prometheus-2.44.0-r1
prometheus-2.44.0-r2
prometheus-2.44.0-r3
prometheus-2.45.0-r0
prometheus-alertmanager-0.25.0-r0
prometheus-alertmanager-0.25.0-r1
prometheus-alertmanager-0.25.0-r2
prometheus-alertmanager-0.25.0-r3
prometheus-alertmanager-0.25.0-r4
prometheus-bind-exporter-0.6.1-r0
prometheus-bind-exporter-0.6.1-r1
prometheus-bind-exporter-0.6.1-r2
prometheus-config-reloader-0.64.0-r0
prometheus-config-reloader-0.64.0-r1
prometheus-config-reloader-0.64.1-r0
prometheus-config-reloader-0.65.0-r0
prometheus-config-reloader-0.65.0-r1
prometheus-config-reloader-0.65.1-r0
prometheus-config-reloader-0.65.1-r1
prometheus-config-reloader-0.65.2-r0
prometheus-config-reloader-0.65.2-r1
prometheus-config-reloader-0.66.0-r0
prometheus-config-reloader-oci-entrypoint-compat-0.65.1-r1
prometheus-config-reloader-oci-entrypoint-compat-0.65.2-r0
prometheus-config-reloader-oci-entrypoint-compat-0.65.2-r1
prometheus-config-reloader-oci-entrypoint-compat-0.66.0-r0
prometheus-elasticsearch-exporter-1.5.0-r0
prometheus-elasticsearch-exporter-1.5.0-r1
prometheus-elasticsearch-exporter-1.5.0-r2
prometheus-elasticsearch-exporter-1.5.0-r3
prometheus-elasticsearch-exporter-1.5.0-r4
prometheus-elasticsearch-exporter-1.6.0-r0
prometheus-mysqld-exporter-0.14.0-r0
prometheus-mysqld-exporter-0.14.0-r1
prometheus-mysqld-exporter-0.14.0-r2
prometheus-mysqld-exporter-0.14.0-r3
prometheus-mysqld-exporter-0.14.0-r4
prometheus-mysqld-exporter-0.14.0-r5
prometheus-mysqld-exporter-0.15.0-r0
prometheus-nats-exporter-0.11.0-r0
prometheus-nats-exporter-0.11.0-r1
prometheus-nats-exporter-0.12.0-r0
prometheus-node-exporter-1.5.0-r0
prometheus-node-exporter-1.5.0-r1
prometheus-node-exporter-1.5.0-r2
prometheus-node-exporter-1.5.0-r3
prometheus-node-exporter-1.5.0-r4
prometheus-node-exporter-1.6.0-r0
prometheus-node-exporter-1.6.0-r1
prometheus-operator-0.64.0-r0
prometheus-operator-0.64.0-r1
prometheus-operator-0.64.1-r0
prometheus-operator-0.65.0-r0
prometheus-operator-0.65.0-r1
prometheus-operator-0.65.1-r0
prometheus-operator-0.65.2-r0
prometheus-operator-0.65.2-r1
prometheus-operator-0.65.2-r2
prometheus-operator-0.66.0-r0
prometheus-postgres-exporter-0.12.0-r0
prometheus-postgres-exporter-0.12.0-r1
prometheus-postgres-exporter-0.12.0-r2
prometheus-postgres-exporter-0.12.0-r3
prometheus-postgres-exporter-0.12.1-r0
prometheus-postgres-exporter-0.13.0-r0
prometheus-postgres-exporter-0.13.0-r1
prometheus-postgres-exporter-0.13.1-r0
prometheus-redis-exporter-1.50.0-r0
prometheus-redis-exporter-1.50.0-r1
prometheus-redis-exporter-1.51.0-r0
prometheus-stackdriver-exporter-0.13.0-r0
prometheus-stackdriver-exporter-0.13.0-r1
prometheus-stackdriver-exporter-0.14.1-r0
prometheus-stackdriver-exporter-0.14.1-r1

# apk add  prometheus=2.41.0-r0
```

to get a specific older version. note that these older versions are "unmaintained" and can/will have vulnerabilities associated with them, but they're still available here.

The same format should work in an apko.yaml if you're using that!

**Reply by henrysachs** [reply]
yeah i'm using apko and have currently only this: 

```
  packages:
    - prometheus
```

**Reply by dlorenc** [reply]
Did sticking on the =$version part work?

We don't have a UI for finding the package versions yet but we're working on it. APK search is probably the easiest way now, but you can also fetch the index directly

**Reply by henrysachs** [reply]
```
  packages:
    - prometheus=2.4.1
```

did not work but with your apk search it seems like they need a bit more than the 3 numbers 😅

**Reply by dlorenc** [reply]
Yeah I think you'll need the epoch part (the -r0)

**Reply by henrysachs** [reply]
Hi @dlorenc sorry for the late response but with the epoch part this works, thanks a lot for this! But can you elaborate a bit on the "unmaintained" part? Chainguard images also contain a tagged version for prometheus for example and I thought they would use the same APKs? Of course if there aren't new APKs of Prometheus itself Prometheus will have vulnerabilities but this is a thing wolfi cant fix because its in the source code of Prometheus isn't it? 
(I read somewhere that you are also able to do some kind of patching for certain packages? It looked like a Git Patch)

**Reply by dlorenc** [reply]
Thanks for the question! As part of the free catalog, we're only planning to maintain one version of each package at a time.

That means patching it for vulnerabilities, and rebuilding the final container image to pick up any of these patches, or any patches in other dependencies in the final image.

In the case of Prometheus, we maintain one version of it here. We'll patch CVEs in Prometheus itself and any runtime dependencies for the final image, but when a new version comes out we update to that.

For other things like say, Java or Postgres we might need to keep multiple versions around - but we treat those as separate packages (java11 and java17, vs java-11.0.0-r0).

So if you build with apko and select older package versions, you're effectively pinning to those. That means you won't automatically get updates.
